Correct all world height validations.

tommysanterre opened the pull request:
Addesses issue https://github.com/mc-server/MCServer/issues/1606.

No more assertion triggered.

madmaxoft commented on src/BlockArea.cpp:
```diff
@@ -441,10 +441,10 @@ bool cBlockArea::Read(cForEachChunkProvider * a_ForEachChunkProvider, int a_MinB
 	{
 		LOGWARNING("%s: MaxBlockY more than chunk height, adjusting to chunk height", __FUNCTION__);
-		a_MaxBlockY = cChunkDef::Height;
+		a_MaxBlockY = cChunkDef::Height - 1;
```
This breaks stuff. `a_MaxBlockY` is supposed to be "one block past the maximum" at this point (see lines 423 - 425), so with your change there's no way to read the top of the chunk

tommysanterre replied:
Thank you, I didn't notice this. I could rename the variables to make it clearer we're working with intervals. Maybe using iterators like naming.
Example: 
    a_MinBlockX becomes a_BeginBlockX
    a_MaxBlockX becomes a_EndBlockX

madmaxoft replied:
No, I don't like that naming, it doesn't convey that the coords need to be sorted, and it indicates no inclusion either. Additionally, the original parameters are inclusive, but the maxima are made exclusive within the function, so either way you name them, one instance will be wrong. I prefer the interface part to be right.

tommysanterre replied:
Oh I see that's the goal of that block. Make them exclusive.

	// Include the Max coords:
	a_MaxBlockX += 1;
	a_MaxBlockY += 1;
	a_MaxBlockZ += 1;

My proposition doesn't make sense then, sorry.

madmaxoft commented on src/Generating/ChunkDesc.cpp:
```diff
@@ -336,10 +336,10 @@ void cChunkDesc::ReadBlockArea(cBlockArea & a_Dest, int a_MinRelX, int a_MaxRelX
 	{
 		LOGWARNING("%s: MaxRelY more than chunk height, adjusting to chunk height", __FUNCTION__);
-		a_MaxRelY = cChunkDef::Height;
+		a_MaxRelY = cChunkDef::Height - 1;
```
Again, the max is supposed to be "one past", so the original test was correct.

tommysanterre replied:
I could rename the variable here as mentioned above.

madmaxoft commented on src/Generating/FinishGen.cpp:
```diff
@@ -124,9 +124,9 @@ void cFinishGenNetherClumpFoliage::TryPlaceClump(cChunkDesc & a_ChunkDesc, int a
+	if (MaxY >= cChunkDef::Height)  // Check if the coordinate is outside the chunk. If it it then adjust it.
 	{
-		MaxY = cChunkDef::Height;
+		MaxY = cChunkDef::Height - 1;
```
The original code was correct, `MaxY` is used as an non-inclusive bound in the for loop below.

tommysanterre replied:
And here as well.

madmaxoft commented on src/Generating/RoughRavines.cpp:
```diff
@@ -139,7 +139,7 @@ class cRoughRavine :
 			rnd = rnd / 4;
 			float Val = (float)(rnd % 256) / 128 - 1;  // Random float in range [-1, +1]
-			if (h + NumBlocks > cChunkDef::Height)
+			if (h + NumBlocks >= cChunkDef::Height)
```
Wrong change.

madmaxoft commented on src/World.cpp:
```diff
@@ -533,7 +533,7 @@ void cWorld::InitializeSpawn(void)
 {
 	m_SpawnX = 0;
-	m_SpawnY = cChunkDef::Height;
+	m_SpawnY = cChunkDef::Height - 1;
```
Nope. We don't want to spawn *in* the terrain, but above it.

madmaxoft commented on src/BlockArea.cpp:
```diff
@@ -441,7 +441,7 @@ bool cBlockArea::Read(cForEachChunkProvider * a_ForEachChunkProvider, int a_MinB
 		a_MaxBlockY = 0;
 	}
-	else if (a_MaxBlockY > cChunkDef::Height)
+	else if (a_MaxBlockY >= cChunkDef::Height)
```
I still think this is a wrong change, we need to be able to read areas with full chunk height (`a_MaxBlockY` is exclusive at this point due to line 425)

tommysanterre replied:
You're right.

madmaxoft commented on src/BlockArea.cpp:
```diff
@@ -517,7 +517,7 @@ bool cBlockArea::Write(cForEachChunkProvider * a_ForEachChunkProvider, int a_Min
 		a_MinBlockY = 0;
 	}
-	else if (a_MinBlockY > cChunkDef::Height - m_Size.y)
+	else if (a_MinBlockY >= cChunkDef::Height - m_Size.y)
```
I'm not sure about this change either; if it is indeed needed, then the assignment on line 523 is wrong.

tommysanterre replied:
Here I thought it was possible to call FastSetBlock with an invalid value but I can't reproduce it even when entering incorrect values with the debugger. I will revert this change also.

tommysanterre commented:
@madmaxoft I really appreciate the reviews you have done for me. I've revised my changes once more and corrected some mistake I also fixed the issue highlighted by your last review. Could you review the code once more?

madmaxoft commented:
Thanks for all the hard work :)
